src/puppetlabs/trapperkeeper/services/webserver/jetty9_websockets.clj

```diff
@@ -0,0 +1,105 @@
+(ns puppetlabs.trapperkeeper.services.webserver.jetty9-websockets
+  (:import (clojure.lang IFn)
+           (org.eclipse.jetty.server Request)
+           (org.eclipse.jetty.websocket.api WebSocketAdapter Session)
+           (org.eclipse.jetty.websocket.server WebSocketHandler)
+           (org.eclipse.jetty.websocket.servlet WebSocketServletFactory WebSocketCreator)
+           (java.security.cert X509Certificate)
+           (java.nio ByteBuffer))
+
+  (:require [clojure.tools.logging :as log]
+            [puppetlabs.websockets.client :refer [WebSocketProtocol]]
+            [schema.core :as schema]))
+
+(def WebsocketHandlers
+  {(schema/optional-key :on-connect) IFn
+   (schema/optional-key :on-error) IFn
+   (schema/optional-key :on-close) IFn
+   (schema/optional-key :on-text) IFn
+   (schema/optional-key :on-bytes) IFn})
+
+(defprotocol WebSocketSend
+  (-send! [x ws] "How to encode content sent to the WebSocket clients"))
+
+(extend-protocol WebSocketSend
+  (Class/forName "[B")
+  (-send! [ba ws]
+    (-send! (ByteBuffer/wrap ba) ws))
+
+  ByteBuffer
+  (-send! [bb ws]
+    (-> ^WebSocketAdapter ws .getRemote (.sendBytes ^ByteBuffer bb)))
+
+  String
+  (-send! [s ws]
+    (-> ^WebSocketAdapter ws .getRemote (.sendString ^String s))))
+
+(extend-protocol WebSocketProtocol
+  WebSocketAdapter
+  (send! [this msg]
+    (-send! msg this))
+  (close! [this]
+    (.. this (getSession) (close)))
+  (remote-addr [this]
+    (.. this (getSession) (getRemoteAddress)))
+  (ssl? [this]
+    (.. this (getSession) (getUpgradeRequest) (isSecure)))
+  (peer-certs [this]
+    (.. this (getCerts)))
+  (idle-timeout! [this ms]
+    (.. this (getSession) (setIdleTimeout ^long ms)))
+  (connected? [this]
+    (. this (isConnected))))
+
+(defn- do-nothing [& args])
+
+(definterface CertGetter
+  (^Object getCerts []))
+
+(schema/defn ^:always-validate proxy-ws-adapter :- WebSocketAdapter
+  [handlers :- WebsocketHandlers
+   x509certs :- [X509Certificate]]
+  (let [{:keys [on-connect on-error on-text on-close on-bytes]
+         :or {on-connect do-nothing
+              on-error do-nothing
+              on-text do-nothing
+              on-close do-nothing
+              on-bytes do-nothing}} handlers]
+    (proxy [WebSocketAdapter CertGetter] []
+      (onWebSocketConnect [^Session session]
+        (let [^WebSocketAdapter this this]
+          (proxy-super onWebSocketConnect session))
+        (on-connect this))
+      (onWebSocketError [^Throwable e]
+        (on-error this e))
+      (onWebSocketText [^String message]
+        (on-text this message))
+      (onWebSocketClose [statusCode ^String reason]
+        (let [^WebSocketAdapter this this]
+          (proxy-super onWebSocketClose statusCode reason))
```

Comment:
The `WebsocketAdapter` implementation of those methods are empty bodies:  https://github.com/eclipse/jetty.project/blob/master/jetty-websocket/websocket-api/src/main/java/org/eclipse/jetty/websocket/api/WebSocketAdapter.java#L54-L80  so for consistency we could dispatch to them, but it seems needless.